src/Microsoft.ML.Transforms/WhiteningTransform.cs

```diff
@@ -207,139 +241,229 @@ private static VersionInfo GetVersionInfo()
                 loaderAssemblyName: typeof(WhiteningTransform).Assembly.FullName);
         }
 
-        private readonly ColInfoEx[] _exes;
-
-        private const string RegistrationName = "Whitening";
+        private readonly ColInfo[] _infos;
 
         /// <summary>
-        /// Convenience constructor for public facing API.
+        /// Initializes a new <see cref="WhiteningTransform"/> object.
         /// </summary>
         /// <param name="env">Host Environment.</param>
-        /// <param name="input">Input <see cref="IDataView"/>. This is the output from previous transform or loader.</param>
-        /// <param name="name">Name of the output column.</param>
-        /// <param name="source">Name of the column to be transformed. If this is null '<paramref name="name"/>' will be used.</param>
-        /// <param name="kind">Whitening kind (PCA/ZCA).</param>
-        public WhiteningTransform(IHostEnvironment env,
-            IDataView input,
-            string name,
-            string source = null,
-            WhiteningKind kind = Defaults.Kind)
-                : this(env, new Arguments() { Column = new[] { new Column() { Source = source ?? name, Name = name } }, Kind = kind }, input)
-        {
-        }
-
-        /// <summary>
-        /// Public constructor corresponding to SignatureDataTransform.
-        /// </summary>
-        public WhiteningTransform(IHostEnvironment env, Arguments args, IDataView input)
-            : base(env, RegistrationName, Contracts.CheckRef(args, nameof(args)).Column,
-                input, TestColumn)
+        /// <param name="inputData">Input data.</param>
+        /// <param name="columns">Describes the settings for the transformer.</param>
+        internal WhiteningTransform(IHostEnvironment env, IDataView inputData, params ColInfo[] columns)
+            : base(Contracts.CheckRef(env, nameof(env)).Register(nameof(WhiteningTransform)), GetColumnPairs(columns))
         {
-            Host.AssertNonEmpty(Infos);
-            Host.Assert(Infos.Length == Utils.Size(args.Column));
+            Host.AssertNonEmpty(ColumnPairs);
+            var inputSchema = inputData.Schema;
 
-            _exes = new ColInfoEx[Infos.Length];
-            for (int i = 0; i < _exes.Length; i++)
-                _exes[i] = new ColInfoEx(args.Column[i], args, Infos[i]);
+            _cols = new int[ColumnPairs.Length];
+            _srcTypes = new ColumnType[ColumnPairs.Length];
+            _infos = columns;
+            for (int i = 0; i < _infos.Length; i++)
+            {
+                if (!inputSchema.TryGetColumnIndex(ColumnPairs[i].input, out _cols[i]))
+                    throw Host.ExceptSchemaMismatch(nameof(inputSchema), "input", ColumnPairs[i].input);
```

Comment:
You already specify `nameof(inputSchema`. Is `"input"` still needed?

Reply:
Unfortunately the parameters to the exception include "column role", so I think I need to keep it, unless you have another suggestion.

---
In reply to: [229562643](https://github.com/dotnet/machinelearning/pull/1452#discussion_r229562643) [](ancestors = 229562643)